### PR TITLE
URL Input: Use Popover instead of custom positionning and custom modals

### LINKED ISSUE #3941: Editor: Anchor's popover displays partially off the screen
## Issue Overview

I spent some time testing Gutenberg version 1.9.0 on my mobile device. This is a follow-up for #3197.

## Steps to Reproduce (for bugs)

Open Gutenberg on iPhone 6s in Safari, iOS 11.0.3.

## Expected Behavior

Everything should be displayed properly on mobile.

## Current Behavior

Sometimes it is displayed partially behind visible area on the left, sometimes on the right.

Another thing is that I would prefer to see the icon to remove the link as the last item. I removed 2 or 3 times link because I somehow assumed it should be the very far item on the list :)

image 5

### PR DESCRIPTION
**Pull request by youknowriad**
This PR updates the URL Input and the format toolbar to use the `Popover` component to display the link modal. The Popover contain behavior to adapt the width and the position to the available space which will fix a lot of issues we have with link modals on mobile.

**Testing instructions**

 - Test the link inputs in the several places we use them (format toolbar, inline format toolbar, image block toolbar, button block...)

closes #3941 #4447 #6183 #6393 #6810

### CODE REVIEW

**Comment by jasmussen** [comment]
> closes #3941 #4447 #6183 #6393 #6810

 🤘

This is a big improvement.

I wish we could, like TinyMCE does, show a little shadow of what's selected in the text, even if your text focus is elsewhere. But that's for separate polish. 

One thing: this doesn't seem to work when the toolbar is fixed to the top.

**Comment by youknowriad** [comment]
> One thing: this doesn't seem to work when the toolbar is fixed to the top.

Good catch, I'll take a look

**Comment by jasmussen** [comment]
> One thing: this doesn't seem to work when the toolbar is fixed to the top.

Actually, scratch that. I don't know what it was that I was seeing, but a hard reload made this work. Stellar work!

Can you move the popover upwards about 10px? that would make it feel more connected with the link.

**Review by jasmussen — APPROVED** [review]
After the latest fix, this is cool from a design perspective.

File: editor/components/rich-text/index.js [hunk 1]
```diff
@@ -446,11 +446,10 @@ export class RichText extends Component {
 	getFocusPosition( position ) {
 		// The container is relatively positioned.
 		const containerPosition = this.containerRef.current.getBoundingClientRect();
-		const toolbarOffset = { top: 10, left: 0 };
 
 		return {
-			top: position.top - containerPosition.top + ( position.height ) + toolbarOffset.top,
-			left: position.left - containerPosition.left + ( position.width / 2 ) + toolbarOffset.left,
+			top: position.top - containerPosition.top + ( position.height ),
```
Comment (iseulde) [inline editor/components/rich-text/index.js]: Tiny thing: unsure why there are brackets here, but they were here before.

**Comment by iseulde** [comment]
Seeing the following error when adding the target attribute, and I don't see this error in master?

```
index.js:292 Uncaught Error: Invalid assumption: expected selection to be within the autocomplete container
    at Autocomplete.getCursor (index.js:292)
    at Autocomplete.search (index.js:443)
    at HTMLUnknownElement.callCallback (react-dom.cca5b42c.js:140)
    at Object.invokeGuardedCallbackDev (react-dom.cca5b42c.js:178)
    at Object.invokeGuardedCallback (react-dom.cca5b42c.js:227)
    at Object.invokeGuardedCallbackAndCatchFirstError (react-dom.cca5b42c.js:241)
    at executeDispatch (react-dom.cca5b42c.js:604)
    at executeDispatchesInOrder (react-dom.cca5b42c.js:626)
    at executeDispatchesAndRelease (react-dom.cca5b42c.js:724)
    at executeDispatchesAndReleaseTopLevel (react-dom.cca5b42c.js:735)
```

**Comment by youknowriad** [comment]
@iseulde I'm seeing the same issue on master

**Comment by noisysocks** [comment]
> I wish we could, like TinyMCE does, show a little shadow of what's selected in the text, even if your text focus is elsewhere. But that's for separate polish.

There's an issue for that: https://github.com/WordPress/gutenberg/issues/6156

**Review by noisysocks — DISMISSED** [review]
I love this ❤️ 

I see just one visual regression. Before there was a nice focus outline on the inline link input:

screen shot 2018-05-24 at 09 59 22

But now that's gone:

screen shot 2018-05-24 at 10 01 23

File: components/popover/index.js [hunk 2]
```diff
@@ -219,6 +220,7 @@ class Popover extends Component {
 			'is-' + xAxis,
 			{
 				'is-mobile': isMobile,
+				'no-arrow': noArrow,
```
Comment (aduth) [inline components/popover/index.js]: This is not very consistent with [modifier class naming](https://github.com/WordPress/gutenberg/blob/master/docs/reference/coding-guidelines.md#naming). I might have preferred something like `has-no-arrow` or `is-without-arrow` or `is-no-arrow`.

Comment (youknowriad) [reply components/popover/index.js]: I like how you see a comment from Andrew from time to time on old PRs :P 

Anyway, I just pushed a fix to master (accidentally, I thought I was on a branch 😬 )

Comment (youknowriad) [reply components/popover/index.js]: https://github.com/WordPress/gutenberg/commit/3ce3feea8b0f2c81df3aee82e8a450466345b786

Comment (aduth) [reply components/popover/index.js]: I enjoy to surf the `git blame` 😄

**Pull request merged**